crates/turbopack-ecmascript/src/references/external_module.rs

```diff
@@ -0,0 +1,226 @@
+use std::io::Write;
+
+use anyhow::{Context, Result};
+use turbo_tasks::Vc;
+use turbo_tasks_fs::{rope::RopeBuilder, FileContent, FileSystem, VirtualFileSystem};
+use turbopack_core::{
+    asset::{Asset, AssetContent},
+    chunk::{AsyncModuleInfo, ChunkItem, ChunkType, ChunkableModule, ChunkingContext},
+    ident::AssetIdent,
+    module::Module,
+    reference::ModuleReferences,
+    resolve::ExternalType,
+};
+
+use crate::{
+    chunk::{
+        EcmascriptChunkItem, EcmascriptChunkItemContent, EcmascriptChunkPlaceable,
+        EcmascriptChunkType, EcmascriptChunkingContext, EcmascriptExports,
+    },
+    references::async_module::{AsyncModule, OptionAsyncModule},
+    utils::StringifyJs,
+    EcmascriptModuleContent,
+};
+
+#[turbo_tasks::function]
+fn layer() -> Vc<String> {
+    Vc::cell("external".to_string())
+}
+
+#[turbo_tasks::value]
+pub struct ExternalModuleAsset {
+    pub request: String,
+    pub external_type: ExternalType,
+    pub import_externals: bool,
+}
+
+impl ExternalModuleAsset {
+    pub fn is_async(&self) -> bool {
+        self.external_type == ExternalType::EcmaScriptModule && self.import_externals
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl ExternalModuleAsset {
+    #[turbo_tasks::function]
+    pub fn new(request: String, external_type: ExternalType, import_externals: bool) -> Vc<Self> {
+        Self::cell(ExternalModuleAsset {
+            request,
+            external_type,
+            import_externals,
+        })
+    }
+
+    #[turbo_tasks::function]
+    pub fn is_async_vc(&self) -> Vc<bool> {
+        Vc::cell(self.is_async())
+    }
+
+    #[turbo_tasks::function]
+    pub fn content(&self) -> Vc<EcmascriptModuleContent> {
+        let mut code = RopeBuilder::default();
+
+        if self.is_async() {
+            writeln!(
+                code,
+                "const mod = await __turbopack_external_import__({});",
+                StringifyJs(&self.request)
+            )
+            .unwrap();
+        } else {
+            writeln!(
+                code,
+                "const mod = __turbopack_external_require__({});",
+                StringifyJs(&self.request)
+            )
+            .unwrap();
+        }
+
+        writeln!(code).unwrap();
+        writeln!(code, "__turbopack_export_namespace__(mod);").unwrap();
+
+        EcmascriptModuleContent {
+            inner_code: code.build(),
+            source_map: None,
+            is_esm: true,
+        }
+        .cell()
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl Module for ExternalModuleAsset {
+    #[turbo_tasks::function]
+    fn ident(&self) -> Vc<AssetIdent> {
+        let fs = VirtualFileSystem::new_with_name("external-assets".to_string());
```

Comment:
```suggestion
        let fs = VirtualFileSystem::new_with_name("externals".to_string());
```

`assets` is internal and we shouldn't expose that use users. The filesystem name is visible in the dev tools